### ISSUE TITLE
when write ganesha hang v5.7

### ISSUE THREAD

**Issue by JYang1986** [nfs-ganesha/nfs-ganesha#1102]
fio -directory=/mnt/ufs_nfs/nfs_fio_test -iodepth=16 -direct=1 -rw=randwrite -ioengine=libaio -bs=4k -numjobs=16 -group_reporting -name=8G_4k_direct -size=8G -offset=0 -time_based -runtime=30000000 -eta=always -eta-newline=1 -eta-interval=1 -thread=1

nfsd-ganesha stack 
when fsal_fd->io_work=0   fsal_fd->fd_work=1 
this thread not be wake up, signal to other wait fd_work thread, but fd_work != 0, still cond_wait, then hang
(gdb) t 25
[Switching to thread 25 (Thread 0x7ffa421ff700 (LWP 116391))]
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
(gdb) bt
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffa5517b42d in reopen_fsal_fd (obj_hdl=0x7ffa43e148c0, openflags=2, fsal_fd=0x7ffa43e14990, can_start=false) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2030
#2  0x00007ffa5517cbd2 in wait_to_start_io (obj_hdl=0x7ffa43e148c0, fsal_fd=0x7ffa43e14990, openflags=2, may_open=true, may_reopen=true) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2383
#3  0x00007ffa5517cf78 in fsal_start_global_io (out_fd=0x7ffa421fd0a8, obj_hdl=0x7ffa43e148c0, my_fd=0x7ffa43e14990, tmp_fd=0x7ffa421fd0b0, openflags=2, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2484
#4  0x00007ffa5517d734 in fsal_start_io (out_fd=0x7ffa421fd0a8, obj_hdl=0x7ffa43e148c0, obj_fd=0x7ffa43e14990, tmp_fd=0x7ffa421fd0b0, state=0x0, openflags=2, open_for_locks=false, reusing_open_state_fd=0x0, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2766
#5  0x00007ffa4d1ef016 in vfs_write2 (obj_hdl=0x7ffa43e148c0, bypass=true, done_cb=0x7ffa552bbd4e <mdc_write_cb>, write_arg=0x7ffa41c3e020, caller_arg=0x7ffa41c29020) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/FSAL_VFS/file.c:1366
#6  0x00007ffa552bbec7 in mdcache_write2 (obj_hdl=0x7ffa43e3d438, bypass=true, done_cb=0x7ffa552a39e5 <nfs3_write_cb>, write_arg=0x7ffa41c3e020, caller_arg=0x7ffa41c3e000) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/Stackable_FSALs/FSAL_MDCACHE/mdcache_file.c:710
#7  0x00007ffa552a4233 in nfs3_write (arg=0x7ffa41c1f830, req=0x7ffa41c1f000, res=0x7ffa3b02d280) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/Protocols/NFS/nfs3_write.c:379
#8  0x00007ffa551b2062 in nfs_rpc_process_request (reqdata=0x7ffa41c1f000, retry=false, t=0x0) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/MainNFSD/nfs_worker_thread.c:1560
#9  0x00007ffa551b252a in nfs_rpc_process_request_slowio (reqdata=0x7ffa41c1f000, retry=false, slowio_check_cb=0x7ffa552dde1a <nfs3_timeout_proc>) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/MainNFSD/nfs_worker_thread.c:1737
#10 0x00007ffa551b2769 in nfs_rpc_valid_NFS (req=0x7ffa41c1f000) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/MainNFSD/nfs_worker_thread.c:1810
#11 0x00007ffa54f3590a in svc_vc_decode (req=0x7ffa41c1f000) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/libntirpc/src/svc_vc.c:1044
#12 0x00007ffa54f30e2d in svc_request (xprt=0x7ffa4960c000, xdrs=0x7ffa41c0c000) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/libntirpc/src/svc_rqst.c:1202
#13 0x00007ffa54f3580f in svc_vc_recv (xprt=0x7ffa4960c000) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/libntirpc/src/svc_vc.c:1017
#14 0x00007ffa54f30dad in svc_rqst_xprt_task_recv (wpe=0x7ffa4960c2c8) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/libntirpc/src/svc_rqst.c:1183
#15 0x00007ffa54f3193c in svc_rqst_epoll_loop (wpe=0x7ffa4f94b118) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/libntirpc/src/svc_rqst.c:1564
#16 0x00007ffa54f3ce03 in work_pool_thread (arg=0x7ffa42200000) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/libntirpc/src/work_pool.c:184
#17 0x00007ffa546902de in start_thread () from /lib64/libpthread.so.0
#18 0x00007ffa53f9fe83 in clone () from /lib64/libc.so.6
(gdb) f 1
#1  0x00007ffa5517b42d in reopen_fsal_fd (obj_hdl=0x7ffa43e148c0, openflags=2, fsal_fd=0x7ffa43e14990, can_start=false) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2030
2030	/export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c: No such file or directory.
(gdb) p *fsal_fd
$1 = {
  openflags = 2,
  fd_work = 1,
  io_work = 0,
  want_read = 0,
  want_write = 12,
  fsal_export = 0x7ffa4f860480,
  fd_lru = {
    next = 0x7ffa43e147f0,
    prev = 0x7ffa55576f40 <fsal_fd_global_lru>
  },
  work_mutex = {
    __data = {
      __lock = 0,
      __count = 0,
      __owner = 0,
      __nusers = 13,
      __kind = 3,
      __spins = 27,
      __elision = 0,
      __list = {
        __prev = 0x0,
        __next = 0x0
      }
    },
    __size = '\000' <repeats 12 times>, "\r\000\000\000\003\000\000\000\033", '\000' <repeats 18 times>,
    __align = 0
  },
  work_cond = {
    __data = {
      {
        __wseq = 39,
        __wseq32 = {
          __low = 39,
          __high = 0
        }
      },
      {
        __g1_start = 1,
        __g1_start32 = {
          __low = 1,
          __high = 0
        }
      },
      __g_refs = {10, 16},
      __g_size = {5, 0},
      __g1_orig_size = 44,
      __wrefs = 104,
      __g_signals = {0, 0}
    },
    __size = "'\000\000\000\000\000\000\000\001\000\000\000\000\000\000\000\n\000\000\000\020\000\000\000\005\000\000\000\000\000\000\000,\000\000\000h\000\000\000\000\000\000\000\000\000\000",
    __align = 39
  },
  close_on_complete = false,
  lru_reclaim = 0,
  fd_type = FSAL_FD_GLOBAL
}
(gdb)

 grep -i2 "fsal_fd=0x7ffa43e14990" nfsd.stack
Thread 42 (Thread 0x7ffa3c1fd700 (LWP 116409)):
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffa5517bf3b in wait_to_start_io (obj_hdl=0x7ffa43e148c0, fsal_fd=0x7ffa43e14990, openflags=2, may_open=true, may_reopen=true) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2242
#2  0x00007ffa5517cf78 in fsal_start_global_io (out_fd=0x7ffa3c1fb0a8, obj_hdl=0x7ffa43e148c0, my_fd=0x7ffa43e14990, tmp_fd=0x7ffa3c1fb0b0, openflags=2, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2484
#3  0x00007ffa5517d734 in fsal_start_io (out_fd=0x7ffa3c1fb0a8, obj_hdl=0x7ffa43e148c0, obj_fd=0x7ffa43e14990, tmp_fd=0x7ffa3c1fb0b0, state=0x0, openflags=2, open_for_locks=false, reusing_open_state_fd=0x0, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2766

Thread 41 (Thread 0x7ffa3c2fe700 (LWP 116408)):
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffa5517bf3b in wait_to_start_io (obj_hdl=0x7ffa43e148c0, fsal_fd=0x7ffa43e14990, openflags=2, may_open=true, may_reopen=true) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2242
#2  0x00007ffa5517cf78 in fsal_start_global_io (out_fd=0x7ffa3c2fc0a8, obj_hdl=0x7ffa43e148c0, my_fd=0x7ffa43e14990, tmp_fd=0x7ffa3c2fc0b0, openflags=2, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2484
#3  0x00007ffa5517d734 in fsal_start_io (out_fd=0x7ffa3c2fc0a8, obj_hdl=0x7ffa43e148c0, obj_fd=0x7ffa43e14990, tmp_fd=0x7ffa3c2fc0b0, state=0x0, openflags=2, open_for_locks=false, reusing_open_state_fd=0x0, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2766

Thread 39 (Thread 0x7ffa3cdff700 (LWP 116406)):
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffa5517bf3b in wait_to_start_io (obj_hdl=0x7ffa43e148c0, fsal_fd=0x7ffa43e14990, openflags=2, may_open=true, may_reopen=true) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2242
#2  0x00007ffa5517cf78 in fsal_start_global_io (out_fd=0x7ffa3cdfd0a8, obj_hdl=0x7ffa43e148c0, my_fd=0x7ffa43e14990, tmp_fd=0x7ffa3cdfd0b0, openflags=2, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2484
#3  0x00007ffa5517d734 in fsal_start_io (out_fd=0x7ffa3cdfd0a8, obj_hdl=0x7ffa43e148c0, obj_fd=0x7ffa43e14990, tmp_fd=0x7ffa3cdfd0b0, state=0x0, openflags=2, open_for_locks=false, reusing_open_state_fd=0x0, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2766

Thread 35 (Thread 0x7ffa3e5ff700 (LWP 116402)):
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffa5517bf3b in wait_to_start_io (obj_hdl=0x7ffa43e148c0, fsal_fd=0x7ffa43e14990, openflags=2, may_open=true, may_reopen=true) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2242
#2  0x00007ffa5517cf78 in fsal_start_global_io (out_fd=0x7ffa3e5fd0a8, obj_hdl=0x7ffa43e148c0, my_fd=0x7ffa43e14990, tmp_fd=0x7ffa3e5fd0b0, openflags=2, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2484
#3  0x00007ffa5517d734 in fsal_start_io (out_fd=0x7ffa3e5fd0a8, obj_hdl=0x7ffa43e148c0, obj_fd=0x7ffa43e14990, tmp_fd=0x7ffa3e5fd0b0, state=0x0, openflags=2, open_for_locks=false, reusing_open_state_fd=0x0, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2766

Thread 32 (Thread 0x7ffa3f7ff700 (LWP 116399)):
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffa5517bf3b in wait_to_start_io (obj_hdl=0x7ffa43e148c0, fsal_fd=0x7ffa43e14990, openflags=2, may_open=true, may_reopen=true) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2242
#2  0x00007ffa5517cf78 in fsal_start_global_io (out_fd=0x7ffa3f7fd0a8, obj_hdl=0x7ffa43e148c0, my_fd=0x7ffa43e14990, tmp_fd=0x7ffa3f7fd0b0, openflags=2, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2484
#3  0x00007ffa5517d734 in fsal_start_io (out_fd=0x7ffa3f7fd0a8, obj_hdl=0x7ffa43e148c0, obj_fd=0x7ffa43e14990, tmp_fd=0x7ffa3f7fd0b0, state=0x0, openflags=2, open_for_locks=false, reusing_open_state_fd=0x0, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2766

Thread 31 (Thread 0x7ffa403ff700 (LWP 116398)):
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffa5517bf3b in wait_to_start_io (obj_hdl=0x7ffa43e148c0, fsal_fd=0x7ffa43e14990, openflags=2, may_open=true, may_reopen=true) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2242
#2  0x00007ffa5517cf78 in fsal_start_global_io (out_fd=0x7ffa403fd0a8, obj_hdl=0x7ffa43e148c0, my_fd=0x7ffa43e14990, tmp_fd=0x7ffa403fd0b0, openflags=2, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2484
#3  0x00007ffa5517d734 in fsal_start_io (out_fd=0x7ffa403fd0a8, obj_hdl=0x7ffa43e148c0, obj_fd=0x7ffa43e14990, tmp_fd=0x7ffa403fd0b0, state=0x0, openflags=2, open_for_locks=false, reusing_open_state_fd=0x0, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2766

Thread 30 (Thread 0x7ffa40dfd700 (LWP 116397)):
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffa5517bf3b in wait_to_start_io (obj_hdl=0x7ffa43e148c0, fsal_fd=0x7ffa43e14990, openflags=2, may_open=true, may_reopen=true) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2242
#2  0x00007ffa5517cf78 in fsal_start_global_io (out_fd=0x7ffa40dfb0a8, obj_hdl=0x7ffa43e148c0, my_fd=0x7ffa43e14990, tmp_fd=0x7ffa40dfb0b0, openflags=2, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2484
#3  0x00007ffa5517d734 in fsal_start_io (out_fd=0x7ffa40dfb0a8, obj_hdl=0x7ffa43e148c0, obj_fd=0x7ffa43e14990, tmp_fd=0x7ffa40dfb0b0, state=0x0, openflags=2, open_for_locks=false, reusing_open_state_fd=0x0, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2766

Thread 29 (Thread 0x7ffa40efe700 (LWP 116396)):
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffa5517bf3b in wait_to_start_io (obj_hdl=0x7ffa43e148c0, fsal_fd=0x7ffa43e14990, openflags=2, may_open=true, may_reopen=true) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2242
#2  0x00007ffa5517cf78 in fsal_start_global_io (out_fd=0x7ffa40efc0a8, obj_hdl=0x7ffa43e148c0, my_fd=0x7ffa43e14990, tmp_fd=0x7ffa40efc0b0, openflags=2, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2484
#3  0x00007ffa5517d734 in fsal_start_io (out_fd=0x7ffa40efc0a8, obj_hdl=0x7ffa43e148c0, obj_fd=0x7ffa43e14990, tmp_fd=0x7ffa40efc0b0, state=0x0, openflags=2, open_for_locks=false, reusing_open_state_fd=0x0, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2766

Thread 28 (Thread 0x7ffa40fff700 (LWP 116395)):
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffa5517bf3b in wait_to_start_io (obj_hdl=0x7ffa43e148c0, fsal_fd=0x7ffa43e14990, openflags=2, may_open=true, may_reopen=true) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2242
#2  0x00007ffa5517cf78 in fsal_start_global_io (out_fd=0x7ffa40ffd0a8, obj_hdl=0x7ffa43e148c0, my_fd=0x7ffa43e14990, tmp_fd=0x7ffa40ffd0b0, openflags=2, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2484
#3  0x00007ffa5517d734 in fsal_start_io (out_fd=0x7ffa40ffd0a8, obj_hdl=0x7ffa43e148c0, obj_fd=0x7ffa43e14990, tmp_fd=0x7ffa40ffd0b0, state=0x0, openflags=2, open_for_locks=false, reusing_open_state_fd=0x0, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2766

Thread 26 (Thread 0x7ffa41bff700 (LWP 116392)):
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffa5517bf3b in wait_to_start_io (obj_hdl=0x7ffa43e148c0, fsal_fd=0x7ffa43e14990, openflags=2, may_open=true, may_reopen=true) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2242
#2  0x00007ffa5517cf78 in fsal_start_global_io (out_fd=0x7ffa41bfd0a8, obj_hdl=0x7ffa43e148c0, my_fd=0x7ffa43e14990, tmp_fd=0x7ffa41bfd0b0, openflags=2, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2484
#3  0x00007ffa5517d734 in fsal_start_io (out_fd=0x7ffa41bfd0a8, obj_hdl=0x7ffa43e148c0, obj_fd=0x7ffa43e14990, tmp_fd=0x7ffa41bfd0b0, state=0x0, openflags=2, open_for_locks=false, reusing_open_state_fd=0x0, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2766

Thread 25 (Thread 0x7ffa421ff700 (LWP 116391)):
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffa5517b42d in reopen_fsal_fd (obj_hdl=0x7ffa43e148c0, openflags=2, fsal_fd=0x7ffa43e14990, can_start=false) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2030
#2  0x00007ffa5517cbd2 in wait_to_start_io (obj_hdl=0x7ffa43e148c0, fsal_fd=0x7ffa43e14990, openflags=2, may_open=true, may_reopen=true) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2383
#3  0x00007ffa5517cf78 in fsal_start_global_io (out_fd=0x7ffa421fd0a8, obj_hdl=0x7ffa43e148c0, my_fd=0x7ffa43e14990, tmp_fd=0x7ffa421fd0b0, openflags=2, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2484
#4  0x00007ffa5517d734 in fsal_start_io (out_fd=0x7ffa421fd0a8, obj_hdl=0x7ffa43e148c0, obj_fd=0x7ffa43e14990, tmp_fd=0x7ffa421fd0b0, state=0x0, openflags=2, open_for_locks=false, reusing_open_state_fd=0x0, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2766

Thread 10 (Thread 0x7ffa489ff700 (LWP 115898)):
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffa5517bf3b in wait_to_start_io (obj_hdl=0x7ffa43e148c0, fsal_fd=0x7ffa43e14990, openflags=2, may_open=true, may_reopen=true) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2242
#2  0x00007ffa5517cf78 in fsal_start_global_io (out_fd=0x7ffa489fd0a8, obj_hdl=0x7ffa43e148c0, my_fd=0x7ffa43e14990, tmp_fd=0x7ffa489fd0b0, openflags=2, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2484
#3  0x00007ffa5517d734 in fsal_start_io (out_fd=0x7ffa489fd0a8, obj_hdl=0x7ffa43e148c0, obj_fd=0x7ffa43e14990, tmp_fd=0x7ffa489fd0b0, state=0x0, openflags=2, open_for_locks=false, reusing_open_state_fd=0x0, bypass=true, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2766

Thread 8 (Thread 0x7ffa495ff700 (LWP 115896)):
#0  0x00007ffa5469648c in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffa5517bf3b in wait_to_start_io (obj_hdl=0x7ffa43e148c0, fsal_fd=0x7ffa43e14990, openflags=32, may_open=false, may_reopen=false) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2242
#2  0x00007ffa5517cf78 in fsal_start_global_io (out_fd=0x7ffa495fce48, obj_hdl=0x7ffa43e148c0, my_fd=0x7ffa43e14990, tmp_fd=0x7ffa495fce50, openflags=32, bypass=false,share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2484
#3  0x00007ffa5517d734 in fsal_start_io (out_fd=0x7ffa495fce48, obj_hdl=0x7ffa43e148c0, obj_fd=0x7ffa43e14990, tmp_fd=0x7ffa495fce50, state=0x0, openflags=32, open_for_locks=false, reusing_open_state_fd=0x0, bypass=false, share=0x7ffa43e14978) at /export/Develop/jy/kun-lun-project/kunlun-ufs/nfs-export/nfs-ganesha/src/FSAL/commonlib.c:2766

**Comment by JYang1986** [comment]
[nfsd_task.txt](https://github.com/nfs-ganesha/nfs-ganesha/files/14681323/nfsd_task.txt)

**Comment by JYang1986** [comment]
This issue may be similar to the following scenario
![nfs-ganesh-hang](https://github.com/nfs-ganesha/nfs-ganesha/assets/22851362/5ff94d98-546a-4005-b3e8-ef327fca2617)

**Comment by ffilz** [comment]
Ah, I think I see the fix. If we use pthread_cond_broadcast, thread 3 will wake up both thread 1 and thread 2. Thread 2 of course goes right back to sleep, but thread 1 can now proceed.

**Comment by JYang1986** [comment]
@ffilz 
 pthread_cond_broadcast maybe not the best method，will wake up threads that should not be wakened
could use two conds io_work_cond, fd_work_cond， like following
![nfs-ganesh-two-cond](https://github.com/nfs-ganesha/nfs-ganesha/assets/22851362/dbee0427-902a-47df-99a7-68e34eb48afc)

**Comment by rojingeorge** [comment]
@JYang1986 - Thanks for the excellent analysis...

I wanted to hunt, why the hang happened in the first place... I could see a lot of contenders for the global fd, tracked by the fsal fd structure - 0x7ffa43e14990, means on the same file many IOs were happening, as suggested by the want_write counters...

This could have been seen, as we missed to decrement fd_work while reclaiming if FD is busy.
When fd_lru starts to clean up the global FDs and finds it to be busy while closing, it was not marking fd_work as complete. Thereby leading to counters never getting decremented & impacting IOs reusing the same fsal_fds by dead waiting on wait_to_start_io.

Please pick up my patch, this would unblock you...This CR decrements the fd_work counter in such case...
https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/1174224

**Comment by JYang1986** [comment]
@rojingeorge

**Comment by JYang1986** [comment]
@rojingeorge 
this patch https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/1174224 not resole this issue
i pick up this patch, still have this issue.
thread 1 write fd_work = 0, it will inc fd_work and enter reopen_fsal_fd， not call fsal_start_fd_work inc fd_work
when you change wait_to_start_io and reopen_fsal_fd LogFullDebug to LogWarn and set log level to warn
this issue may be easier to reproduce

**Comment by ffilz** [comment]
Yea, your analysis shows that we needed thread 1 to wakeup when signaled by thread 3 instead of thread 2. For a quick and dirty solution, and to validate the analysis, could you change the signal to a broadcast. Then thread 1 and thread 2 will be woken up and I think it will unjam.

The real fix I think does need to be two condition variables, one of which we will have to broadcast to allow I/O waiters to all proceed.

**Comment by Haroldll** [comment]
Is there any progress on this issue? 
I also found the same problem，the vdbench script I use is:
```
hd=default,vdbench=/home/vdbench,user=root,shell=ssh
hd=hd1,system=localhost
fsd=fsd1,anchor=/test1,depth=1,width=1,files=400,sizes=10g,shared=yes
fwd=format,threads=128,xfersize=4k
fwd=default,xfersize=4k,fileio=random,fileselect=random,rdpct=60,threads=128
fwd=fwd1,fsd=fsd*, host=hd1
rd=rd1,fwd=fwd*,fwdrate=max,format=restart,elapsed=400000,interval=1
```
./vdbench -f test -vr

**Comment by ffilz** [comment]
Hmm, so the mutex and condition variable are global. Separating won't necessarily save anything as we still have to broadcast because there might be threads waiting to do fd work on multiple different files. I will propose moving the mutex and condition variables into the fsal_fd

**Comment by ffilz** [comment]
Oops, never mind, wrong condition variable, there already is one per fsal_fd...

**Comment by ffilz** [comment]
I haven't had time to test it but a fix:

https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/1188188?usp=search

There are 2 patches, introduction of PTHREAD_COND_broadcast is in a separate patch.

**Comment by JYang1986** [comment]
@ffilz 
PTHREAD_COND_broadcast not define
and PTHREAD_COND_signal log use Wait cond info

**Comment by ffilz** [comment]
> @ffilz PTHREAD_COND_broadcast not define 

You also need the related patch: https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/1188187?usp=search

> PTHREAD_COND_signal log use Wait cond info

I can fix that up in the related patch

**Comment by ffilz** [comment]
Updated the related patch and changed the names of the condition variables for clarity.

**Comment by ffilz** [comment]
Issue #1063 is marked as a duplicate of this and the patch resolves that issue.

**Comment by JYang1986** [comment]
@ffilz 
The patch you posted on solves the issue for me。
